fix 9213 DNU when browsing implementors

hogoww opened the pull request:
fixes #9213.

This breaks (at least) the test SPCodePresenterTest >>  #testSelectedSelector.
I left it as is for now and I'm awaiting for some feedback / opinion on this.
The breaking assert in the test is not relevant IMO in this test.
I'll see if the CI finds more breaking tests and their relevance.

The problem is two fold.
First, when reproducing the issue, the string is parse (rightfully) as a Variable (an unresolved one), which is put in a doIt (#noMethod).
The selector extraction is looking for a selector and only encounters a variable.
Therefore, the first selector it's actually encountering is the doIt selector #noMethod.
There is a test showing it in the selector extraction &  the fix only in the doIt case.
I am unsure whether this can happen some other way.

The second problem is that when putting the caret at the end of a selector, it's too big for the string
'isTest' size = 6
Position of cursor when at the end = 7.
Therefore it was never extracted as a selector.
Fix without test. 
Unsure where this should go if anywhere.

Thanks for reading, awaiting for CI & feedbacks :)

hogoww committed:
[fix] fixes #9213

hogoww commented:
As far as my manual testing go, this is consistent with expected behavior, I think

MarcusDenker commented:
SpCodePresenterTest>>testSelectedSelector fails:
```Error
Got #anObjectdo instead of #noMethod.
Stacktrace
TestFailure
Got #anObjectdo instead of #noMethod.
SpCodePresenterTest(TestAsserter)>>assert:description:resumable:
SpCodePresenterTest(TestAsserter)>>assert:description:
SpCodePresenterTest(TestAsserter)>>assert:equals:
SpCodePresenterTest>>testSelectedSelector
SpCodePresenterTest(TestCase)>>performTest
```

hogoww commented:
> SpCodePresenterTest>>testSelectedSelector fails:
> 
> ```
> Got #anObjectdo instead of #noMethod.
> Stacktrace
> TestFailure
> Got #anObjectdo instead of #noMethod.
> SpCodePresenterTest(TestAsserter)>>assert:description:resumable:
> SpCodePresenterTest(TestAsserter)>>assert:description:
> SpCodePresenterTest(TestAsserter)>>assert:equals:
> SpCodePresenterTest>>testSelectedSelector
> SpCodePresenterTest(TestCase)>>performTest
> ```

See my comment about that in previous post =)

hogoww committed:
[improvement] add more tests, takes into account when you have empty strings, or faulty cases

hogoww committed:
[improvement] SpCodePresenter behavior testing of selector has been extracted into the CNSelectorExtractionOnSelectionTest. The only one remaining is the one about the nil text, which apparently happens.

hogoww committed:
[fix] recover tests erased by mistake

hogoww commented:
I extracted the logic of the spec test into the Extractor tests.
I left the nil case which is irrelevant for extractor tests.
I added more tests, and am now treating more cases.
Should be ready for integration, if the CI says it's okay :)

I'm of course available for changes if you spot anything !

estebanlm commented:
so, what´s the next step of this? 
for me the change is good, we can adapt `testSelectedSelector` in consequence.

estebanlm commented on src/Spec2-Code/SpCodePresenter.class.st:
```diff
@@ -197,7 +197,8 @@ SpCodePresenter >> browseSelectedSelectorIfGlobal: globalBlock ifNotGlobal: nonG
 
 	variableOrClassName := self selectedSelector.
-
+	variableOrClassName ifNil: [ ^ self inform: 'No string could be parsed to answer your query.' ].
```
in a second thought I do not like this. 
what to show if there are no results (or showing it at all) is a responsibility of the users of this.

hogoww replied:
I think that's a good point.
However, given how the messages are done here, I'm very unsure about putting an "extra case" in the sender.
doBrowseSenders/Implementors just sends a message with two blocks and does not expect a result.

That would look like:
doBrowseSenders

    res := self
        browseSelectedSelectorIfGlobal: [ :global | 
            self systemNavigation browseAllSendersOf: global ]
        ifNotGlobal:  [ :selector |
            self systemNavigation browseAllReferencesTo: selector ].
       res ifNotNil: [  self inform: blablabla ]

I feel that's not better.
Moreover it creates a bit of code duplication for the two senders of this message.
Also, this is kind of a UI method already, isn't it?

estebanlm reviewed (changes requested):
In second thoughts, I prefer is you do not call directly the UI from that method. 
If needed, add a new keyword to the message ifNone: or something, so the users can choose what to do (probably the same as now, but in the right place :P)

estebanlm commented on src/Tools-CodeNavigation/CNSelectorExtractor.class.st:
```diff
@@ -80,6 +80,11 @@ CNSelectorExtractor >> extractSelectorFromNode: aNode [
 			node := node parent messages first ] ].
+	
+	"In the case of doIt methods"
+	node selector = #noMethod & aNode isVariable
```
not 
```Smalltalk
node selector = #noMethod and: [ aNode isVariable ]
```
?

hogoww replied:
Looks like it matters very little in this case.
I can change it if you'd like =)

estebanlm reviewed (commented):
yes I am ok with that :) 
(I need to learn how this thing of changes work, I do not even know where this comment will arise :P)

estebanlm commented:
yes, please... is about good practices more than immediate consequences :)

On May 6 2021, at 2:03 pm, hogoww ***@***.***> wrote:
>
>
> @hogoww commented on this pull request.
> In src/Tools-CodeNavigation/CNSelectorExtractor.class.st (https://github.com/pharo-project/pharo/pull/9220#discussion_r627356199):
>
> > @@ -80,6 +80,11 @@ CNSelectorExtractor >> extractSelectorFromNode: aNode [
> (node ~= originalNode and: [
> node parent notNil and: [ node parent isCascade ] ]) ifTrue: [
> node := node parent messages first ] ].
> +
> + "In the case of doIt methods"
> + node selector = #noMethod & aNode isVariable
> Looks like it matters very little in this case.
> I can change it if you'd like =)
>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub (https://github.com/pharo-project/pharo/pull/9220#discussion_r627356199), or unsubscribe (https://github.com/notifications/unsubscribe-auth/AAD5MXVW5M5IUBUYID2RPU3TMKASFANCNFSM4333XNYQ).
>

hogoww committed:
[fix] requested change

hogoww commented:
Done !

hogoww commented:
Failing tests look unrelated

estebanlm commented:
what are we waiting to integrate this?

hogoww commented:
One "requested change was not tagged as resolved although it was.
Apart from that, I don't know.
I don't intend to change anything else so far :)

MarcusDenker commented:
I downloaded and tested the build artefact: fixes the problem
